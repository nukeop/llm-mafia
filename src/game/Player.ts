export enum Team {
  Machines,
  Humans,
}

export class Player {
  constructor(
    public name: string,
    public team: Team,
  ) {}
}
