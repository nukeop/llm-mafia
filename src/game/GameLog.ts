import { ChatCompletionMessageParam } from 'openai/resources';
import { Player } from './Player';

export enum ActionType {
  Thought = 'thought',
  Speech = 'speech',
  Vote = 'vote',
  EndTurn = 'end_turn',
}

export enum MessageType {
  System = 'system',
  Error = 'error',
  PlayerAction = 'player_action',
  Announcer = 'announcer',
}

type PlayerAction = {
  player: Player;
  content: string;
  actionType: ActionType;
  type: MessageType.PlayerAction;

  // OpenAI API call ID
  callId?: string;
};

type SystemMessage = {
  content: string;
  type: MessageType.System;
};

type ErrorMessage = {
  content: string;
  type: MessageType.Error;
};

type AnnouncerMessage = {
  content: string;
  type: MessageType.Announcer;
};

export type LogMessage =
  | PlayerAction
  | SystemMessage
  | ErrorMessage
  | AnnouncerMessage;

export class GameLog {
  #messages: LogMessage[] = [];

  constructor() {
    this.#messages = [];
  }

  get messages(): LogMessage[] {
    return this.#messages;
  }

  addPlayerAction(
    player: Player,
    content: string,
    actionType: ActionType,
    callId?: string,
  ) {
    this.#messages.push({
      player,
      content,
      type: MessageType.PlayerAction,
      actionType,
      callId,
    });
  }

  addSystemMessage(content: string) {
    this.#messages.push({ content, type: MessageType.System });
  }

  addErrorMessage(content: string) {
    this.#messages.push({ content, type: MessageType.Error });
  }

  addAnnouncerMessage(content: string) {
    this.#messages.push({ content, type: MessageType.Announcer });
  }

  formatLogForLLM(player: Player): ChatCompletionMessageParam[] {
    const messages = this.#messages.filter(
      (message) =>
        isSpeech(message) ||
        isAnnouncerMessage(message) ||
        (isThought(message) && (message as PlayerAction).player === player),
    ) as (PlayerAction | AnnouncerMessage)[];

    return messages.reduce((result: ChatCompletionMessageParam[], message) => {
      if (isAnnouncerMessage(message)) {
        result.push({
          role: 'user',
          content: `[Announcer}: ${message.content}`,
        });
      } else if (message.actionType === ActionType.Thought) {
        result.push({
          role: 'assistant',
          tool_calls: [
            {
              id: message.callId ?? '',
              type: 'function',
              function: {
                name: ActionType.Thought,
                arguments: JSON.stringify({ content: message.content }),
              },
            },
          ],
        });

        result.push({
          role: 'tool',
          tool_call_id: message.callId ?? '',
          content: 'Success',
        });
      } else {
        if (message.player === player) {
          result.push({
            role: 'assistant',
            tool_calls: [
              {
                id: message.callId ?? '',
                type: 'function',
                function: {
                  name: ActionType.Speech,
                  arguments: JSON.stringify({ content: message.content }),
                },
              },
            ],
          });
          result.push({
            role: 'tool',
            tool_call_id: message.callId ?? '',
            content: 'Success',
          });
        } else {
          result.push({
            role: 'user',
            content: `[${message.player.name}]: ${message.content}`,
          });
        }
      }
      return result;
    }, []);
  }

  formatLogForGameEnd(): string {
    return this.#messages
      .map((message) => {
        if (isSystemMessage(message)) {
          return `[System]: ${message.content}`;
        } else if (isPlayerAction(message)) {
          if (message.actionType === ActionType.EndTurn) {
            return `*${message.player.name} ends his turn*`;
          }

          if (message.actionType === ActionType.Vote) {
            return `*${message.player.name} votes for ${message.content}*`;
          }

          return `[${message.player.name}](${actionTypeToString(message.actionType)}): ${message.content}`;
        } else if (isAnnouncerMessage(message)) {
          return `[Announcer]: ${message.content}`;
        } else {
          return message.content;
        }
      })
      .join('\n');
  }
}

export function isPlayerAction(message: LogMessage): message is PlayerAction {
  return message.type === MessageType.PlayerAction;
}

export function isSystemMessage(message: LogMessage): message is SystemMessage {
  return message.type === MessageType.System;
}

export function isErrorMessage(message: LogMessage): message is ErrorMessage {
  return message.type === MessageType.Error;
}

export function isAnnouncerMessage(
  message: LogMessage,
): message is AnnouncerMessage {
  return message.type === MessageType.Announcer;
}

export function isSpeech(message: LogMessage): boolean {
  return isPlayerAction(message) && message.actionType === ActionType.Speech;
}

export function isThought(message: LogMessage): boolean {
  return isPlayerAction(message) && message.actionType === ActionType.Thought;
}

export function actionTypeToString(actionType: ActionType): string {
  switch (actionType) {
    case ActionType.Thought:
      return 'Thought';
    case ActionType.Speech:
      return 'Speech';
    case ActionType.Vote:
      return 'Vote';
    case ActionType.EndTurn:
      return 'End Turn';
    default:
      return 'Unknown';
  }
}
