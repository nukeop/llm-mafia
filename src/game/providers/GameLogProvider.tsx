import React, { createContext, useContext, useState } from 'react';
import { ChatCompletionMessageParam } from 'openai/resources';
import { Player } from '../Player';
import FileLogger from '../../file-logger';

export enum ActionType {
  Thought = 'thought',
  Speech = 'speech',
  Vote = 'vote',
  EndTurn = 'end_turn',
}

export enum MessageType {
  System = 'system',
  Error = 'error',
  PlayerAction = 'player_action',
  Announcer = 'announcer',
}

export type PlayerAction = {
  player: Player;
  content: string;
  actionType: ActionType;
  type: MessageType.PlayerAction;
  reason?: string;

  // OpenAI API call ID
  callId?: string;
};

export type SystemMessage = {
  content: string;
  type: MessageType.System;
};

export type ErrorMessage = {
  content: string;
  type: MessageType.Error;
};

export type AnnouncerMessage = {
  content: string;
  type: MessageType.Announcer;
};

export type LogMessage =
  | PlayerAction
  | SystemMessage
  | ErrorMessage
  | AnnouncerMessage;

export type GameLogContextValue = {
  messages: LogMessage[];
  addPlayerAction: (
    player: Player,
    content: string,
    actionType: ActionType,
    callId?: string,
    reason?: string,
  ) => void;
  addSystemMessage: (content: string) => void;
  addErrorMessage: (content: string) => void;
  addAnnouncerMessage: (content: string) => void;
  formatLogForLLM: (player: Player) => ChatCompletionMessageParam[];
};

const GameLogContext = createContext<GameLogContextValue | undefined>(
  undefined,
);

export const useGameLog = (): GameLogContextValue => {
  const context = useContext(GameLogContext);
  if (!context) {
    throw new Error('useGameLog must be used within a GameLogProvider');
  }
  return context;
};

type GameLogProviderProps = {
  children: React.ReactNode;
};
export const GameLogProvider: React.FC<GameLogProviderProps> = ({
  children,
}) => {
  const [messages, setMessages] = useState<LogMessage[]>([]);

  const addPlayerAction = (
    player: Player,
    content: string,
    actionType: ActionType,
    callId?: string,
    reason?: string,
  ) => {
    FileLogger.info(playerActionToString(player, content, actionType));
    setMessages((prevMessages) => [
      ...prevMessages,
      {
        player,
        content,
        type: MessageType.PlayerAction,
        actionType,
        callId,
        reason,
      },
    ]);
  };

  const addSystemMessage = (content: string) => {
    FileLogger.info(`[System]: ${content}`);
    setMessages((prevMessages) => [
      ...prevMessages,
      { content, type: MessageType.System },
    ]);
  };

  const addErrorMessage = (content: string) => {
    FileLogger.error(`[Error]: ${content}`);
    setMessages((prevMessages) => [
      ...prevMessages,
      { content, type: MessageType.Error },
    ]);
  };

  const addAnnouncerMessage = (content: string) => {
    FileLogger.info(`[Announcer]: ${content}`);
    setMessages((prevMessages) => [
      ...prevMessages,
      { content, type: MessageType.Announcer },
    ]);
  };

  const formatLogForLLM = (player: Player) => {
    const filteredMessages = messages.filter(
      (message) =>
        isSpeech(message) ||
        isAnnouncerMessage(message) ||
        (isThought(message) && (message as PlayerAction).player === player) ||
        (isVote(message) && (message as PlayerAction).player === player) ||
        (isEndTurn(message) && (message as PlayerAction).player === player),
    ) as (PlayerAction | AnnouncerMessage)[];

    return filteredMessages.reduce(
      (result: ChatCompletionMessageParam[], message) => {
        if (isAnnouncerMessage(message)) {
          result.push({
            role: 'user',
            content: `[Announcer}: ${message.content}`,
          });
        } else if (
          message.actionType === ActionType.Thought ||
          message.actionType === ActionType.Vote ||
          message.actionType === ActionType.EndTurn
        ) {
          result.push({
            role: 'assistant',
            tool_calls: [
              {
                id: message.callId ?? '',
                type: 'function',
                function: {
                  name: message.actionType,
                  arguments: JSON.stringify({ content: message.content }),
                },
              },
            ],
          });

          result.push({
            role: 'tool',
            tool_call_id: message.callId ?? '',
            content: 'Success',
          });
        } else {
          if (message.player === player) {
            result.push({
              role: 'assistant',
              tool_calls: [
                {
                  id: message.callId ?? '',
                  type: 'function',
                  function: {
                    name: ActionType.Speech,
                    arguments: JSON.stringify({ content: message.content }),
                  },
                },
              ],
            });
            result.push({
              role: 'tool',
              tool_call_id: message.callId ?? '',
              content: 'Success',
            });
          } else {
            result.push({
              role: 'user',
              content: `[${message.player.name}]: ${message.content}`,
            });
          }
        }
        return result;
      },
      [],
    );
  };

  const playerActionToString = (
    player: Player,
    content: string,
    actionType: ActionType,
  ): string => {
    if (actionType === ActionType.EndTurn) {
      return `*${player.name} ends his turn*`;
    }

    if (actionType === ActionType.Vote) {
      return `*${player.name} votes for ${content}*`;
    }

    return `[${player.name}](${actionTypeToString(actionType)}): ${content}`;
  };

  const value: GameLogContextValue = {
    messages,
    addPlayerAction,
    addSystemMessage,
    addErrorMessage,
    addAnnouncerMessage,
    formatLogForLLM,
  };

  return (
    <GameLogContext.Provider value={value}>{children}</GameLogContext.Provider>
  );
};

export function isPlayerAction(message: LogMessage): message is PlayerAction {
  return message.type === MessageType.PlayerAction;
}

export function isSystemMessage(message: LogMessage): message is SystemMessage {
  return message.type === MessageType.System;
}

export function isErrorMessage(message: LogMessage): message is ErrorMessage {
  return message.type === MessageType.Error;
}

export function isAnnouncerMessage(
  message: LogMessage,
): message is AnnouncerMessage {
  return message.type === MessageType.Announcer;
}

export function isSpeech(message: LogMessage): boolean {
  return isPlayerAction(message) && message.actionType === ActionType.Speech;
}

export function isThought(message: LogMessage): boolean {
  return isPlayerAction(message) && message.actionType === ActionType.Thought;
}

export function isVote(message: LogMessage): boolean {
  return isPlayerAction(message) && message.actionType === ActionType.Vote;
}

export function isEndTurn(message: LogMessage): boolean {
  return isPlayerAction(message) && message.actionType === ActionType.EndTurn;
}

export function actionTypeToString(actionType: ActionType): string {
  switch (actionType) {
    case ActionType.Thought:
      return 'Thought';
    case ActionType.Speech:
      return 'Speech';
    case ActionType.Vote:
      return 'Vote';
    case ActionType.EndTurn:
      return 'End Turn';
    default:
      return 'Unknown';
  }
}
