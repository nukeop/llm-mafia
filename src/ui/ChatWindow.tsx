import React, { useEffect, useState } from 'react';
import { Box, Text, useInput } from 'ink';
import { ActionType, LogMessage, MessageType } from '../game/GameLog';
import { Speech } from './Speech';
import { SystemText } from './SystemText';
import { ErrorText } from './ErrorText';
import { Vote } from './Vote';

type ChatWindowProps = {
  messages: LogMessage[];
  linesToShow: number;
};

export const ChatWindow: React.FC<ChatWindowProps> = ({
  messages,
  linesToShow,
}) => {
  const [scrollPosition, setScrollPosition] = useState(0);
  const [visibleRange, setVisibleRange] = useState({ start: 0, end: linesToShow });

  useEffect(() => {
    setScrollPosition(Math.max(0, messages.length - linesToShow));
    setVisibleRange({ start: Math.max(0, messages.length - linesToShow), end: messages.length });
  }, [messages, linesToShow]);

  useInput((input, key) => {
    if (key.upArrow) {
      setScrollPosition((prev) => Math.max(0, prev - 1));
      setVisibleRange((prev) => ({
        start: Math.max(0, prev.start - 1),
        end: Math.max(linesToShow, prev.end - 1),
      }));
    } else if (key.downArrow) {
      setScrollPosition((prev) =>
        Math.min(messages.length - linesToShow, prev + 1),
      );
      setVisibleRange((prev) => ({
        start: Math.min(messages.length - linesToShow, prev.start + 1),
        end: Math.min(messages.length, prev.end + 1),
      }));
    }
  });

  const visibleMessages = messages.slice(visibleRange.start, visibleRange.end);

  return (
    <Box flexDirection="column" height={linesToShow + 2}>
      <Box
        borderStyle="round"
        padding={1}
        marginBottom={1}
        flexDirection="column"
        alignItems="flex-end"
      >
        {scrollPosition > 0 && <Text color="gray">▲ Scroll up</Text>}
        {scrollPosition < messages.length - linesToShow && (
          <Text color="gray">▼ Scroll down</Text>
        )}
      </Box>
      {visibleMessages.map((message, index) => {
        switch (message.type) {
          case MessageType.PlayerAction:
            if (message.actionType === ActionType.Speech) {
              return (
                <Speech
                  key={index}
                  player={message.player}
                  content={message.content}
                />
              );
            } else if (message.actionType === ActionType.Vote) {
              return (
                <Vote
                  key={index}
                  player={message.player}
                  content={message.content}
                />
              );
            }
            break;
          case MessageType.System:
            return <SystemText key={index} content={message.content} />;
          case MessageType.Error:
            return <ErrorText key={index} content={message.content} />;
          default:
            return null;
        }
      })}
    </Box>
  );
};
