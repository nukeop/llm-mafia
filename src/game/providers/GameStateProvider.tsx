import React, { createContext, useContext, useEffect, useState } from 'react';
import { Player, Team } from '../Player';
import { sample } from 'lodash';
import { ActionType, useGameLog } from './GameLogProvider';
import Logger from '../../logger';
import { names } from '../../prompts/names';
import { personalities } from '../../prompts/personalities';
import { useMachinePlayerAction } from '../hooks/useMachinePlayerAction';

interface GameStateContextType {
  machinePlayers: Player[];
  humanPlayer: Player;
  actingPlayer: Player;
  initGameState: (numberOfPlayers: number) => void;
  advance: () => Promise<void>;
  isHumanTurn: () => boolean;
}

const GameStateContext = createContext<GameStateContextType | undefined>(
  undefined,
);

export const useGameState = () => {
  const context = useContext(GameStateContext);
  if (context === undefined) {
    throw new Error('useGameState must be used within a GameStateProvider');
  }
  return context;
};

type GameStateProviderProps = {
  children: React.ReactNode;
};

export const GameStateProvider: React.FC<GameStateProviderProps> = ({
  children,
}) => {
  const [machinePlayers, setMachinePlayers] = useState<Player[]>([]);
  const [humanPlayer, setHumanPlayer] = useState<Player>(
    new Player('', Team.Humans),
  );
  const [actingPlayer, setActingPlayer] = useState<Player>(humanPlayer);
  const { addSystemMessage, addErrorMessage } = useGameLog();

  const playerNames = (): string[] => {
    return [humanPlayer, ...machinePlayers].map((player) => player.name);
  };

  const { processPlayerAction } = useMachinePlayerAction({
    actingPlayer,
    playerNames: playerNames(),
  });

  useEffect(() => {
    if (!machinePlayers.length || !humanPlayer.name) {
      return;
    }

    addSystemMessage(
      'Welcome to the LLM Mafia! You are the human player. The goal is to blend in while the machines try to eliminate you. Good luck!',
    );
    addSystemMessage(
      `The machine players are:\n ${machinePlayers.map((player) => player.name).join('\n ')}.`,
    );
    addSystemMessage(`You are ${humanPlayer.name}.`);
  }, [machinePlayers, humanPlayer]);

  const initGameState = (numberOfPlayers: number) => {
    let availableNames = [...names];
    const machinePlayersInit = Array.from({ length: numberOfPlayers }, (_) => {
      const name = sample(availableNames) ?? '';
      availableNames = availableNames.filter((n) => n !== name);
      return new Player(name!, Team.Machines, sample(personalities)?.name);
    });

    const humanName = sample(availableNames) ?? '';
    const humanPlayerInit = new Player(humanName!, Team.Humans);

    setMachinePlayers(machinePlayersInit);
    setHumanPlayer(humanPlayerInit);
    setActingPlayer(machinePlayersInit[0]);
  };

  /**
   * Proceed to the next player.
   * If the current player is the last machine player, the next player is the human player.
   * Otherwise, the next player is the next machine player.
   */
  const nextPlayer = () => {
    if (actingPlayer === humanPlayer) {
      setActingPlayer(machinePlayers[0]);
    } else {
      const currentPlayerIndex = machinePlayers.findIndex(
        (player) => player === actingPlayer,
      );

      if (currentPlayerIndex === machinePlayers.length - 1) {
        setActingPlayer(humanPlayer);
      } else {
        const nextPlayerIndex = currentPlayerIndex + 1;
        setActingPlayer(machinePlayers[nextPlayerIndex]);
      }
    }
  };

  const isHumanTurn = (): boolean => {
    return actingPlayer === humanPlayer;
  };

  const advance = async () => {
    if (actingPlayer === humanPlayer) {
      nextPlayer();
    } else {
      try {
        Logger.debug(`Processing machine turn for: ${actingPlayer.name}`);
        const action = await processPlayerAction();

        if (action === ActionType.EndTurn) {
          nextPlayer();
        }
      } catch (error) {
        addErrorMessage('An error occurred while processing the machine turn.');
        addErrorMessage((error as Error).message);
        nextPlayer();
      }
    }
  };

  return (
    <GameStateContext.Provider
      value={{
        machinePlayers,
        humanPlayer,
        actingPlayer,
        initGameState,
        advance,
        isHumanTurn,
      }}
    >
      {children}
    </GameStateContext.Provider>
  );
};
