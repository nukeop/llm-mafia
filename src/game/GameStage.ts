import { GameState } from './GameState';
import { Player } from './Player';

export class GameStage {
  #gameState?: GameState;
  #actingPlayer: Player;

  constructor(actingPlayer: Player) {
    this.#actingPlayer = actingPlayer;
  }

  connectGameState(gameState: GameState): void {
    this.#gameState = gameState;
  }

  get actingPlayer(): Player {
    return this.#actingPlayer;
  }

  /**
   * Proceed to the next player.
   * If the current player is the last machine player, the next player is the human player.
   * @returns The next player.
   * @throws Error if GameStage is not connected to GameState.
   */
  nextPlayer(): Player {
    if (!this.#gameState)
      throw new Error('GameStage not connected to GameState');

    if (this.#actingPlayer === this.#gameState.humanPlayer) {
      this.#actingPlayer = this.#gameState.machinePlayers[0];
    } else {
      const machinePlayers = this.#gameState.machinePlayers;
      const currentPlayerIndex = machinePlayers.findIndex(
        (player) => player === this.#actingPlayer,
      );

      if (currentPlayerIndex === machinePlayers.length - 1) {
        this.#actingPlayer = this.#gameState.humanPlayer;
      } else {
        const nextPlayerIndex = currentPlayerIndex + 1;
        this.#actingPlayer = machinePlayers[nextPlayerIndex];
      }
    }

    return this.#actingPlayer;
  }
}
