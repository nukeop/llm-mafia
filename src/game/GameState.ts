import { sample } from 'lodash';
import { Player, Team } from './Player';
import { names } from '../prompts/names';
import { ActionType, GameLog } from './GameLog';
import { GameStage } from './GameStage';
import { createSystemPrompt } from '../prompts';
import { OpenAiApiService } from '../services/OpenAiService';
import { tools } from '../prompts/tools';
import Logger from '../logger';
import { safeParse } from '../json-utils';
import { personalities } from '../prompts/personalities';

export class GameState {
  #machinePlayers: Player[];
  #humanPlayer: Player;
  #log: GameLog;
  #stage: GameStage;

  constructor(machinePlayers: Player[], humanPlayer: Player) {
    this.#machinePlayers = machinePlayers;
    this.#humanPlayer = humanPlayer;
    this.#log = new GameLog();
    this.#stage = new GameStage(this.machinePlayers[0]);
  }

  get machinePlayers(): Player[] {
    return this.#machinePlayers;
  }

  get humanPlayer(): Player {
    return this.#humanPlayer;
  }

  get log(): GameLog {
    return this.#log;
  }

  get stage(): GameStage {
    return this.#stage;
  }

  playerNames(): string[] {
    return [this.humanPlayer, ...this.machinePlayers].map(
      (player) => player.name,
    );
  }

  async processPlayerAction(): Promise<void> {
    const prompt = createSystemPrompt(
      this.stage.actingPlayer.name,
      this.playerNames(),
      this.stage.actingPlayer.personality!,
    );
    const log = this.log.formatLogForLLM(this.stage.actingPlayer);
    Logger.debug(JSON.stringify(log, null, 2));
    const service = new OpenAiApiService();
    const response = await service.createChatCompletion({
      max_tokens: 512,
      model: 'gpt-4o',
      tools,
      parallel_tool_calls: false,
      messages: [{ role: 'system', content: prompt }, ...log],
    });

    const choice = response.choices[0];
    Logger.debug(JSON.stringify(choice, null, 2));
    const toolCall = choice.message?.tool_calls?.[0];
    const message = choice.message?.content;

    if (toolCall) {
      const toolCallBody = safeParse(toolCall?.function.arguments!);
      const action = toolCall?.function.name;
      const actionType: ActionType = action as ActionType;

      if (actionType === ActionType.Speech) {
        this.log.addPlayerAction(
          this.stage.actingPlayer,
          toolCallBody.content,
          ActionType.Speech,
          toolCall.id,
        );
      }

      if (actionType === ActionType.Thought) {
        this.log.addPlayerAction(
          this.stage.actingPlayer,
          toolCallBody.content,
          ActionType.Thought,
          toolCall.id,
        );
      }

      if (actionType === ActionType.Vote) {
        // this.stage.addVote(this.stage.actingPlayer, toolCallBody.target);
      }

      if (actionType === ActionType.EndTurn) {
        return;
      }

      await this.processPlayerAction();
    } else if (message) {
      this.log.addPlayerAction(
        this.stage.actingPlayer,
        message,
        ActionType.Speech,
      );

      await this.processPlayerAction();
    }
  }

  async advance() {
    if (this.stage.actingPlayer === this.humanPlayer) {
      this.stage.nextPlayer();
    } else {
      try {
        Logger.debug(
          `Processing machine turn for: ${this.stage.actingPlayer.name}`,
        );
        await this.processPlayerAction();

        this.stage.nextPlayer();
        Logger.debug(
          `Progressing to next player: ${this.stage.actingPlayer.name}`,
        );
      } catch (error) {
        this.log.addErrorMessage(
          'An error occurred while processing the machine turn.',
        );
        this.log.addErrorMessage((error as Error).message);
        this.stage.nextPlayer();
      }
    }
  }
}

export const initGameState = (numberOfPlayers: number): GameState => {
  let availableNames = names;
  const machinePlayers = Array.from({ length: numberOfPlayers }, (_) => {
    const name = sample(availableNames) ?? '';
    availableNames = availableNames.filter((n) => n !== name);
    return new Player(name!, Team.Machines, sample(personalities)!);
  });

  const humanName = sample(names) ?? '';
  const state = new GameState(
    machinePlayers,
    new Player(humanName!, Team.Humans),
  );
  state.stage.connectGameState(state);
  return state;
};
