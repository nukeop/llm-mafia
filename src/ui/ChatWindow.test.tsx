import { render } from 'ink-testing-library';
import { ChatWindow } from './ChatWindow';
import React from 'react';
import { ActionType, MessageType } from '../game/providers/GameLogProvider';
import { Player, Team } from '../game/Player';

describe('Chat window', () => {
  it('renders an empty chat window', () => {
    const t = render(<ChatWindow messages={[]} />);
    expect(t.lastFrame()).toMatchSnapshot();
  });

  it('renders a chat window with messages', async () => {
    const player = new Player('Mr. Test', Team.Machines);

    const messages = [
      {
        content: 'Announcement',
        type: MessageType.Announcer,
      } as const,
      {
        player,
        content: 'Hello, world!',
        type: MessageType.PlayerAction,
        actionType: ActionType.Speech,
      } as const,
      { content: 'Error', type: MessageType.Error } as const,
      { content: 'System message', type: MessageType.System } as const,
    ];

    const t = render(<ChatWindow messages={messages} />);

    expect(t.lastFrame()).toMatchSnapshot();
  });
});
