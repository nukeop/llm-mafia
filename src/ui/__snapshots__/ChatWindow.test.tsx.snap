// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat window renders a chat window with messages 1`] = `
"â•­â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚                                                                                                  â”‚
â”‚                                                                                                  â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯

[32m[Mr. Test]: [37mHello, world![39m[32m[39m
[31mError[39m

[33mSystem message[39m


"
`;

exports[`Chat window renders an empty chat window 1`] = `
"â•­â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•®
â”‚                                                                                                  â”‚
â”‚                                                                                                  â”‚
â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â•¯







"
`;
